README.md
# Sensu-Puppet

#### Table of Contents

1. [Module Description](#module-description)
2. [Setup - The basics of getting started with sensu](#setup)
    * [What sensu affects](#what-sensu-affects)
    * [Setup requirements](#setup-requirements)
    * [Beginning with sensu](#beginning-with-sensu)
3. [Usage - Configuration options and additional functionality](#usage)
    * [Basic Sensu backend](#basic-sensu-backend)
    * [Basic Sensu agent](#basic-sensu-agent)
    * [Advanced agent](#advanced-agent)
    * [Advanced SSL](#advanced-ssl)
    * [Enterprise support](#enterprise-support)
    * [Installing Plugins](#installing-plugins)
    * [Installing Extensions](#installing-extensions)
    * [Exported resources](#exported-resources)
    * [Resource purging](#resource-purging)
    * [Sensu backend cluster](#sensu-backend-cluster)
        * [Adding backend members to an existing cluster](#adding-backend-members-to-an-existing-cluster)
4. [Reference](#reference)
    * [Facts](#facts)
5. [Limitations - OS compatibility, etc.](#limitations)
6. [Development - Guide for contributing to the module](#development)
7. [License](#license)

## Module description

Installs and manages [Sensu](http://sensuapp.org), the open source monitoring framework.

Please note, that this is a **Partner Supported** module, which means that technical customer support for this module is solely provided by Sensu. Puppet does not provide support for any **Partner Supported** modules. Technical support for this module is provided by Sensu at [https://sensuapp.org/support](https://sensuapp.org/support).

### Documented with Puppet Strings

[Puppet Strings documentation](http://sensu.github.io/sensu-puppet/doc/)

### Compatibility - supported sensu versions

If not explicitly stated it should always support the latest Sensu release.
Please log an issue if you identify any incompatibilities.

| Sensu Version   | Recommended Puppet Module Version   |
| --------------- | ----------------------------------- |
| 5.x             | latest v3 |
| 0.26.0 - 1.x    | latest v2 |
| 0.22.x - 0.25.x | 2.1.0                               |
| 0.20.x - 0.21.x | 2.0.0                               |
| 0.17.x - 0.19.x | 1.5.5                               |


### Upgrade note

Sensu Go 5.x is a rewrite of Sensu and no longer depends on redis and rabbitmq. Version 3 of this module supports Sensu Go 5.x.

## Setup

### What sensu effects

This module will install packages, create configuration and start services necessary to manage Sensu agents and backend.

### Setup requirements

Plugin sync is required if the custom sensu types and providers are used.

This module has a soft dependency on the [puppetlabs/apt](https://forge.puppet.com/puppetlabs/apt) module (`>= 5.0.1 < 7.0.0`) for systems using `apt`.

If using Puppet >= 6.0.0 there is a soft dependency on the [puppetlabs/yumrepo_core](https://forge.puppet.com/puppetlabs/yumrepo_core) module (`>= 1.0.1 < 2.0.0`) for systems using `yum`.

### Beginning with sensu

This module provides Vagrant definitions that can be used to get started with Sensu.

```bash
vagrant up sensu-backend
vagrant ssh sensu-backend
```

#### Beginning with a Sensu cluster

Multiple Vagrant boxes are available for testing a sensu-backend cluster.

```bash
vagrant up sensu-backend-peer1 sensu-backend-peer2
vagrant provision sensu-backend-peer1 sensu-backend-peer2
```

## Usage

### Basic Sensu backend

The following example will configure sensu-backend, sensu-agent on backend and add a check.
By default this module will configure the backend to use Puppet's SSL certificate and CA.
It's advisable to not rely on the default password. Changing the password requires providing the previous password via `old_password`.

```puppet
  class { 'sensugo::backend':
    password     => 'supersecret',
    old_password => 'P@ssw0rd!',
  }
  include sensugo::agent
  sensugo_check { 'check-cpu':
    ensure        => 'present',
    command       => 'check-cpu.sh -w 75 -c 90',
    interval      => 60,
    subscriptions => ['linux'],
  }
```

### Basic Sensu agent

The following example will manage resources necessary to configure a sensu-agent to communicate with a sensu-backend and
associated to `linux` and `apache-servers` subscriptions.

```puppet
  class { 'sensugo::agent':
    backends    => ['sensu-backend.example.com:8081'],
    config_hash => {
      'subscriptions' => ['linux', 'apache-servers'],
    },
  }
```

### Advanced agent

If you wish to change the `agent` password you must provide the new and old password.
It's advisable to set `show_diff` to `false` to avoid exposing the agent password.

```puppet
class { 'sensugo::backend':
  agent_password     => 'supersecret',
  agent_old_password => 'P@ssw0rd!',
}
class { 'sensugo::agent':
  config_hash => {
    'password' => 'supersecret',
  },
  show_diff   => false,
}
```

### Advanced SSL

By default this module uses Puppet's SSL certificates and CA.
If you would prefer to use different certificates override the `ssl_ca_source`, `ssl_cert_source` and `ssl_key_source` parameters.
The value for `url_host` must be valid for the provided certificate and the value used for agent's `backends` must also match the certificate used by the specified backend.
If the certificates and keys are already installed then define the source parameters as filesystem paths.

```puppet
class { 'sensu':
  ssl_ca_source => 'puppet:///modules/profile/sensu/ca.pem',
}
class { 'sensugo::backend':
  url_host        => 'sensu-backend.example.com',
  ssl_cert_source => 'puppet:///modules/profile/sensu/cert.pem',
  ssl_key_source  => 'puppet:///modules/profile/sensu/key.pem',
}
```
```puppet
class { 'sensu':
  ssl_ca_source => 'puppet:///modules/profile/sensu/ca.pem',
}
class { 'sensugo::agent':
  backends    => ['sensu-backend.example.com:8081'],
  config_hash => {
    'subscriptions' => ['linux', 'apache-servers'],
  },
}
```

To disable SSL support:

```puppet
class { 'sensu':
  use_ssl => false,
}
```

### Enterprise Support

In order to activate enterprise support the license file needs to be added:

```puppet
class { 'sensugo::backend':
  license_source => 'puppet:///modules/profile/sensu/license.json',
}
```

The type `sensugo_ldap_auth` requires a valid enterprise license.

### Installing Plugins

Plugin management is handled by the `sensugo::plugins` class.

Example installing plugins on agent:

```puppet
  class { 'sensugo::agent':
    backends    => ['sensu-backend.example.com:8081'],
    config_hash => {
      'subscriptions' => ['linux', 'apache-servers'],
    },
  }
  class { 'sensugo::plugins':
    plugins => ['disk-checks'],
  }
```

The `plugins` parameter can also be a Hash that sets the version:

```puppet
  class { 'sensugo::agent':
    backends    => ['sensu-backend.example.com:8081'],
    config_hash => {
      'subscriptions' => ['linux', 'apache-servers'],
    },
  }
  class { 'sensugo::plugins':
    plugins => {
      'disk-checks' => { 'version' => 'latest' },
    },
  }
```

Set `dependencies` to an empty Array to disable the `sensugo::plugins` dependency management.

```puppet
  class { 'sensugo::plugins':
    dependencies => [],
  }
```

You can uninstall plugins by passing `ensure` as `absent`.

```puppet
  class { 'sensugo::agent':
    backends    => ['sensu-backend.example.com:8081'],
    config_hash => {
      'subscriptions' => ['linux', 'apache-servers'],
    },
  }
  class { 'sensugo::plugins':
    plugins => {
      'disk-checks' => { 'ensure' => 'absent' },
    },
  }
```

### Installing Extensions

Extension management is handled by the `sensugo::plugins` class.

Example installing extension on backend:

```puppet
  class { 'sensugo::backend':
    password     => 'supersecret',
    old_password => 'P@ssw0rd!',
  }
  class { 'sensugo::plugins':
    extensions => ['graphite'],
  }
```

The `extensions` parameter can also be a Hash that sets the version:

```puppet
  class { 'sensugo::backend':
    password     => 'supersecret',
    old_password => 'P@ssw0rd!',
  }
  class { 'sensugo::plugins':
    extensions => {
      'graphite' => { 'version' => 'latest' },
    },
  }
```

You can uninstall extensions by passing `ensure` as `absent`.

```puppet
  class { 'sensugo::backend':
    password     => 'supersecret',
    old_password => 'P@ssw0rd!',
  }
  class { 'sensugo::plugins':
    extensions => {
      'graphite' => { 'ensure' => 'absent' },
    },
  }
```

### Exported resources

One possible approach to defining checks is having agents export their checks to the sensu-backend using [Exported Resources](https://puppet.com/docs/puppet/latest/lang_exported.html).

The following example would be defined for agents:

```puppet
  @@sensugo_check { 'check-cpu':
    ensure        => 'present',
    command       => 'check-cpu.sh -w 75 -c 90',
    interval      => 60,
    subscriptions => ['linux'],
  }
```

The backend system would collect all `sensugo_check` resources.

```puppet
  sensugo_check <<||>>
```

### Resource purging

All the types provided by this module support purging except `sensugo_config`.
At this time `sensugo_asset` can not be purged, see [Limitations](#limitations).
This example will remove all unmanaged Sensu checks:

```puppet
resources { 'sensugo_check':
  purge => true,
}
```

### Sensu backend cluster

A `sensu-backend` cluster can be defined for fresh installs by defining the necessary `config_hash` values.
The following examples are using Hiera and assume the `sensugo::backend` class is included.

```yaml
# data/fqdn/sensu-backend1.example.com.yaml
---
sensugo::backend::config_hash:
  etcd-advertise-client-urls: "http://%{facts.ipaddress}:2379"
  etcd-listen-client-urls: "http://%{facts.ipaddress}:2379"
  etcd-listen-peer-urls: 'http://0.0.0.0:2380'
  etcd-initial-cluster: 'backend1=http://192.168.0.1:2380,backend2=http://192.168.0.2:2380'
  etcd-initial-advertise-peer-urls: "http://%{facts.ipaddress}:2380"
  etcd-initial-cluster-state: 'new'
  etcd-initial-cluster-token: ''
  etcd-name: 'backend1'
```
```yaml
# data/fqdn/sensu-backend2.example.com.yaml
---
sensugo::backend::config_hash:
  etcd-advertise-client-urls: "http://%{facts.ipaddress}:2379"
  etcd-listen-client-urls: "http://%{facts.ipaddress}:2379"
  etcd-listen-peer-urls: 'http://0.0.0.0:2380'
  etcd-initial-cluster: 'backend1=http://192.168.0.1:2380,backend2=http://192.168.0.2:2380'
  etcd-initial-advertise-peer-urls: "http://%{facts.ipaddress}:2380"
  etcd-initial-cluster-state: 'new'
  etcd-initial-cluster-token: ''
  etcd-name: 'backend2'
```

#### Adding backend members to an existing cluster

Adding new members to an existing cluster requires two steps.

First, add the member to the catalog on one of the existing cluster backends with the `sensugo_cluster_member` type.

```puppet
sensugo_cluster_member { 'backend3':
  peer_urls => ['http://192.168.0.3:2380'],
}
```

Second, configure and start `sensu-backend` to interact with the existing cluster.
The output from Puppet when a new `sensugo_cluster_member` is applied will print some of the values needed.

```yaml
# data/fqdn/sensu-backend3.example.com.yaml
---
sensugo::backend::config_hash:
  etcd-advertise-client-urls: "http://%{facts.ipaddress}:2379"
  etcd-listen-client-urls: "http://%{facts.ipaddress}:2379"
  etcd-listen-peer-urls: 'http://0.0.0.0:2380'
  etcd-initial-cluster: 'backend1=http://192.168.0.1:2380,backend2=http://192.168.0.2:2380,backend3=http://192.168.0.3:2380'
  etcd-initial-advertise-peer-urls: "http://%{facts.ipaddress}:2380"
  etcd-initial-cluster-state: 'existing'
  etcd-initial-cluster-token: ''
  etcd-name: 'backend3'
```

The first step will not fully add the node to the cluster until the second step is performed.

## Reference

### Facts

#### `sensugo_agent`

The `sensugo_agent` fact returns the Sensu agent version information by the `sensu-agent` binary.

```shell
facter -p sensugo_agent
{
  version => "5.1.0",
  build => "b2ea9fcdb21e236e6e9a7de12225a6d90c786c57",
  built => "2018-12-18T21:31:11+0000"
}
```

#### `sensugo_backend`

The `sensugo_backend` fact returns the Sensu backend version information by the `sensu-backend` binary.

```shell
facter -p sensugo_backend
{
  version => "5.1.0",
  build => "b2ea9fcdb21e236e6e9a7de12225a6d90c786c57",
  built => "2018-12-18T21:31:11+0000"
}
```

#### `sensuctl`

The `sensuctl` fact returns the sensuctl version information by the `sensuctl` binary.

```shell
facter -p sensuctl
{
  version => "5.1.0",
  build => "b2ea9fcdb21e236e6e9a7de12225a6d90c786c57",
  built => "2018-12-18T21:31:11+0000"
}
```

## Limitations

The Sensu v2 support is designed so that all resources managed by `sensuctl` are defined on the `sensu-backend` host.
This module does not support adding `sensuctl` resources on a host other than the `sensu-backend` host.

The type `sensugo_asset` does not at this time support `ensure => absent` due to a limitation with sensuctl, see [sensu-go#988](https://github.com/sensu/sensu-go/issues/988).

The type `sensugo_user` does not at this time support `ensure => absent` due to a limitation with sensuctl, see [sensu-go#2540](https://github.com/sensu/sensu-go/issues/2540).

### Notes regarding support

This module is built for use with Puppet versions 5 and 6 and the ruby
versions associated with those releases. See `.travis.yml` for an exact
matrix of Puppet releases and ruby versions.

This module targets the latest release of the current major Puppet
version and the previous major version. Platform support will be removed
when a platform is no longer supported by Puppet, Sensu or the platform
maintainer has signaled that it is end of life (EOL).

Though Amazon does not announce end of life (EOL) for its releases, it
does encourage you to use the latest releases. This module will support
the current release and the previous release. Since AWS does not release
Vagrant boxes and the intent of those platforms is to run in AWS, we
will not maintain Vagrant systems for local development for Amazon
Linux.

### Supported Platforms

* EL 6
* EL 7
* Debian 8
* Debian 9
* Ubuntu 14.04 LTS
* Ubuntu 16.04 LTS
* Ubuntu 18.04 LTS
* Amazon 2018.03
* Amazon 2

## Development

See [CONTRIBUTING.md](CONTRIBUTING.md)

## License

See [LICENSE](LICENSE) file.


CHANGELOG.md
# Change Log

## [v2.52.0](https://github.com/sensu/sensu-puppet/tree/v2.52.0)

[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.51.0...v2.52.0)

**Implemented enhancements:**

- Implement self.instances for all native providers [\#770](https://github.com/sensu/sensu-puppet/issues/770)

**Fixed bugs:**

- Windows sensu purge issue [\#879](https://github.com/sensu/sensu-puppet/issues/879)
- Use correct version of stdlib with macos-client in vagrant [\#892](https://github.com/sensu/sensu-puppet/pull/892) ([ghoneycutt](https://github.com/ghoneycutt))

**Closed issues:**

- Vagrant tests fail for macos-client [\#891](https://github.com/sensu/sensu-puppet/issues/891)

**Merged pull requests:**

- add sensu version facter \(continued\) [\#894](https://github.com/sensu/sensu-puppet/pull/894) ([treydock](https://github.com/treydock))

## [v2.51.0](https://github.com/sensu/sensu-puppet/tree/v2.51.0) (2018-04-05)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.50.1...v2.51.0)

**Merged pull requests:**

- Enable specifying enable/ensure when managing services [\#889](https://github.com/sensu/sensu-puppet/pull/889) ([ghoneycutt](https://github.com/ghoneycutt))

## [v2.50.1](https://github.com/sensu/sensu-puppet/tree/v2.50.1) (2018-04-05)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.50.0...v2.50.1)

**Fixed bugs:**

- sensu\_enterprise\_dashboard\_api\_config purge [\#883](https://github.com/sensu/sensu-puppet/issues/883)

**Merged pull requests:**

- \(GH-883\) Remove example sensu\_enterprise\_dashboard\_api\_config resources by default [\#885](https://github.com/sensu/sensu-puppet/pull/885) ([treydock](https://github.com/treydock))

## [v2.50.0](https://github.com/sensu/sensu-puppet/tree/v2.50.0) (2018-01-19)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.49.0...v2.50.0)

**Closed issues:**

- Pull Request: sensugo::check auto\_resolve [\#857](https://github.com/sensu/sensu-puppet/issues/857)
- Enterprise - Purging files no longer managed fails to notify correct service [\#854](https://github.com/sensu/sensu-puppet/issues/854)
- When removing a check, sensu service not refresh [\#782](https://github.com/sensu/sensu-puppet/issues/782)

**Merged pull requests:**

- Adding auto\_resolve param to sensugo::check. Replaces \#858 [\#872](https://github.com/sensu/sensu-puppet/pull/872) ([alvagante](https://github.com/alvagante))
- Restart sensu-enterprise service when configs are purged \#854 [\#871](https://github.com/sensu/sensu-puppet/pull/871) ([alvagante](https://github.com/alvagante))

## [v2.49.0](https://github.com/sensu/sensu-puppet/tree/v2.49.0) (2018-01-16)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.48.0...v2.49.0)

**Implemented enhancements:**

- Allow for management of file and directory permissions [\#825](https://github.com/sensu/sensu-puppet/issues/825)

**Merged pull requests:**

- Add parameters to configure dir and file modes [\#869](https://github.com/sensu/sensu-puppet/pull/869) ([ghoneycutt](https://github.com/ghoneycutt))

## [v2.48.0](https://github.com/sensu/sensu-puppet/tree/v2.48.0) (2018-01-15)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.47.0...v2.48.0)

**Implemented enhancements:**

- Not all Sensu Enterprise Dashboard options are available [\#866](https://github.com/sensu/sensu-puppet/issues/866)
- Add support for Debian 9 stretch [\#708](https://github.com/sensu/sensu-puppet/issues/708)

**Merged pull requests:**

- \[866\] Add Sensu Enterprise Dashboard auth and oidc configuration options [\#867](https://github.com/sensu/sensu-puppet/pull/867) ([treydock](https://github.com/treydock))

## [v2.47.0](https://github.com/sensu/sensu-puppet/tree/v2.47.0) (2018-01-15)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.46.0...v2.47.0)

**Implemented enhancements:**

- \(GH-708\) Add support for Debian 9 \(Stretch\) [\#795](https://github.com/sensu/sensu-puppet/pull/795) ([ghoneycutt](https://github.com/ghoneycutt))

## [v2.46.0](https://github.com/sensu/sensu-puppet/tree/v2.46.0) (2018-01-15)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.45.0...v2.46.0)

**Closed issues:**

- upgrading to newer version of sensu with newer embeded ruby doesn't reinstall plugins [\#542](https://github.com/sensu/sensu-puppet/issues/542)

**Merged pull requests:**

- \[542\] Add dependencies for sensu\_gem plugins [\#817](https://github.com/sensu/sensu-puppet/pull/817) ([glarizza](https://github.com/glarizza))

## [v2.45.0](https://github.com/sensu/sensu-puppet/tree/v2.45.0) (2018-01-09)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.44.0...v2.45.0)

**Closed issues:**

- Support MacOS client [\#862](https://github.com/sensu/sensu-puppet/issues/862)

**Merged pull requests:**

- Macos [\#863](https://github.com/sensu/sensu-puppet/pull/863) ([ghoneycutt](https://github.com/ghoneycutt))

## [v2.44.0](https://github.com/sensu/sensu-puppet/tree/v2.44.0) (2018-01-04)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.43.0...v2.44.0)

**Merged pull requests:**

- Use latest puppetlabs/stdlib \(2.24.0\) and Stdlib::Filemode type [\#865](https://github.com/sensu/sensu-puppet/pull/865) ([ghoneycutt](https://github.com/ghoneycutt))

## [v2.43.0](https://github.com/sensu/sensu-puppet/tree/v2.43.0) (2018-01-04)
[Full Changelog](https://github.com/sensu/sensu-puppet/compare/v2.42.0...v2.43.0)

**Implemented enhancements:**

- MAX\_OPEN\_FILES should be configurable for Sensu Enterprise [\#849](https://github.com/sensu/sensu-puppet/issues/849)

**Fixed bugs:**

- Roundrobin subscriptions on Windows aren't configured [\#820](https://github.com/sensu/sensu-puppet/issues/820)

**Closed issues:**

- check "type" field lost in json file while upgrading module [\#860](https://github.com/sensu/sensu-puppet/issues/860)

lib/puppet/type/sensu_check.rb
require_relative '../../puppet_x/sensu/type'
require_relative '../../puppet_x/sensu/array_property'
require_relative '../../puppet_x/sensu/array_of_hashes_property'
require_relative '../../puppet_x/sensu/hash_property'
require_relative '../../puppet_x/sensu/integer_property'

Puppet::Type.newtype(:sensugo_check) do
  desc <<-DESC
@summary Manages Sensu checks
@example Create a check
  sensugo_check { 'test':
    ensure        => 'present',
    command       => 'check-http.rb',
    subscriptions => ['demo'],
    handlers      => ['email'],
    interval      => 60,
  }

@example Create a check that has a hook
  sensugo_check { 'test':
    ensure        => 'present',
    command       => 'check-cpu.sh -w 75 -c 90',
    subscriptions => ['linux'],
    check_hooks   => [
      { 'critical' => ['ps'] },
      { 'warning'  => ['ps'] },
    ],
    interval      => 60,
  }

**Autorequires**:
* `Package[sensu-go-cli]`
* `Service[sensu-backend]`
* `sensugo_configure[puppet]`
* `sensugo_api_validator[sensu]`
* `sensugo_namespace` - Puppet will autorequire `sensugo_namespace` resource defined in `namespace` property.
* `sensugo_handler` - Puppet will autorequie `sensugo_handler` resources defined in `handlers` property.
* `sensugo_asset` - Puppet will autorequire `sensugo_asset` resources defined in `runtime_assets` property.
* `sensugo_hook` - Puppet will autorequire `sensugo_hook` resources defined in `check_hooks` property.
DESC

  extend PuppetX::Sensugo::Type
  add_autorequires()

  ensurable

  newparam(:name, :namevar => true) do
    desc "The name of the check."
    validate do |value|
      unless value =~ /^[\w\.\-]+$/
        raise ArgumentError, "sensugo_check name invalid"
      end
    end
  end

  newproperty(:command) do
    desc "The check command to be executed."
  end

  newproperty(:subscriptions, :array_matching => :all, :parent => PuppetX::Sensugo::ArrayProperty) do
    desc "An array of Sensu entity subscriptions that check requests will be sent to."
  end

  newproperty(:handlers, :array_matching => :all, :parent => PuppetX::Sensugo::ArrayProperty) do
    desc "An array of Sensu event handlers (names) to use for events created by the check."
  end

  newproperty(:interval, :parent => PuppetX::Sensugo::IntegerProperty) do
    desc "The frequency in seconds the check is executed."
    newvalues(/^[0-9]+$/, :absent)
  end

  newproperty(:cron) do
    desc 'When the check should be executed, using the Cron syntax.'
    newvalues(/.*/, :absent)
  end

  newproperty(:publish, :boolean => true) do
    desc "If check requests are published for the check."
    newvalues(:true, :false)
    defaultto(:true)
  end

  newproperty(:timeout, :parent => PuppetX::Sensugo::IntegerProperty) do
    desc "The check execution duration timeout in seconds (hard stop)."
    newvalues(/^[0-9]+$/, :absent)
  end

  newproperty(:ttl, :parent => PuppetX::Sensugo::IntegerProperty) do
    desc "The time to live (TTL) in seconds until check results are considered stale."
    newvalues(/^[0-9]+$/, :absent)
  end

  newproperty(:stdin, :boolean => true) do
    desc "If the Sensu agent writes JSON serialized Sensu entity and check data to the command process' STDIN"
    newvalues(:true, :false)
    defaultto(:false)
  end

  newproperty(:low_flap_threshold, :parent => PuppetX::Sensugo::IntegerProperty) do
    desc "The flap detection low threshold (% state change) for the check"
    newvalues(/^[0-9]+$/, :absent)
  end

  newproperty(:high_flap_threshold, :parent => PuppetX::Sensugo::IntegerProperty) do
    desc "The flap detection high threshold (% state change) for the check"
    newvalues(/^[0-9]+$/, :absent)
  end

  newproperty(:runtime_assets, :array_matching => :all, :parent => PuppetX::Sensugo::ArrayProperty) do
    desc "An array of Sensu assets (names), required at runtime for the execution of the command"
    newvalues(/.*/, :absent)
  end

  newproperty(:check_hooks, :array_matching => :all, :parent => PuppetX::Sensugo::ArrayOfHashesProperty) do
    desc "An array of check response types with respective arrays of Sensu hook names."
    validate do |value|
      if ! value.is_a?(Hash)
        raise ArgumentError, "check_hooks elements must be a Hash"
      end
      if value.keys.size > 1
        raise ArgumentError, "check_hooks Hash must only contain one key"
      end
      type = value.keys[0]
      hooks = value[type]
      type_valid = false
      if ['ok','warning','critical','unknown','non-zero'].include?(type)
        type_valid = true
      elsif type.to_s =~ /^\d+$/ && type.to_i.between?(0,255)
        type_valid = true
      end
      if ! type_valid
        raise ArgumentError, "check_hooks type #{type} is invalid"
      end
      if ! hooks.is_a?(Array)
        raise ArgumentError, "check_hooks hooks must be an Array"
      end
    end
    munge do |value|
      type = value.keys[0]
      hooks = value[type]
      { type.to_s => hooks }
    end
  end

  newproperty(:proxy_entity_name) do
    desc "The entity name, used to create a proxy entity for an external resource (i.e., a network switch)."
    newvalues(/^[\w\.\-]+$/, :absent)
    validate do |value|
      unless value =~ /^[\w\.\-]+$/
        raise ArgumentError, "check_check proxy_entity_name invalid"
      end
    end
  end

  newproperty(:round_robin, :boolean => true) do
    desc "If the check should be executed on a single entity within a subscription in a round-robin fashion."
    newvalues(:true, :false)
  end

  newproperty(:proxy_requests_entity_attributes, :array_matching => :all, :parent => PuppetX::Sensugo::ArrayProperty) do
    desc "Sensu entity attributes to match entities in the registry, using Sensu Query Expressions"
  end

  newproperty(:proxy_requests_splay, :boolean => true) do
    desc "If proxy check requests should be splayed"
    newvalues(:true, :false)
  end

  newproperty(:proxy_requests_splay_coverage, :parent => PuppetX::Sensugo::IntegerProperty) do
    desc "The splay coverage percentage use for proxy check request splay calculation."
  end

  newproperty(:silenced, :boolean => true) do
    desc "If the event is to be silenced."
    newvalues(:true, :false)
  end

  newproperty(:env_vars, :array_matching => :all, :parent => PuppetX::Sensugo::ArrayProperty) do
    desc "An array of environment variables to use with command execution."
    newvalues(/.*/, :absent)
  end

  newproperty(:output_metric_format) do
    desc "The metric format generated by the check command."
    newvalues(:nagios_perfdata, :graphite_plaintext, :influxdb_line, :opentsdb_line, :absent)
  end

  newproperty(:output_metric_handlers, :array_matching => :all, :parent => PuppetX::Sensugo::ArrayProperty) do
    desc "An array of Sensu handlers to use for events created by the check."
    newvalues(/.*/, :absent)
  end

  newproperty(:max_output_size, :parent => PuppetX::Sensugo::IntegerProperty) do
    desc 'Maximum size, in bytes, of stored check outputs.'
  end

  newproperty(:discard_output, :boolean => true) do
    desc 'Discard check output after extracting metrics.'
    newvalues(:true, :false)
  end

  newproperty(:namespace) do
    desc "The Sensu RBAC namespace that this check belongs to."
    defaultto 'default'
  end

  newproperty(:labels, :parent => PuppetX::Sensugo::HashProperty) do
    desc "Custom attributes to include with event data, which can be queried like regular attributes."
  end

  newproperty(:annotations, :parent => PuppetX::Sensugo::HashProperty) do
    desc "Arbitrary, non-identifying metadata to include with event data."
  end

  autorequire(:sensugo_handler) do
    self[:handlers]
  end

  autorequire(:sensugo_asset) do
    self[:runtime_assets]
  end

  autorequire(:sensugo_hook) do
    check_hooks = []
    (self[:check_hooks] || []).each do |check_hook|
      check_hook.each_pair do |severity, hooks|
        hooks.each do |hook|
          check_hooks << hook
        end
      end
    end
    check_hooks
  end

  validate do
    required_properties = [
      :command,
      :subscriptions,
    ]
    required_properties.each do |property|
      if self[:ensure] == :present && self[property].nil?
        fail "You must provide a #{property}"
      end
    end
    if self[:ensure] == :present
      if (self[:interval].nil? && self[:cron].nil?) && self[:publish] == :true
        fail "interval or cron is required if publish is true"
      end
      if self[:interval].nil? && ! (self[:publish] == :false || ! self[:cron].nil?)
        fail "interval is required unless publish is false or cron is defined"
      end
      if self[:cron].nil? && ! (self[:publish] == :false || ! self[:interval].nil?)
        fail "cron is required unless publish is false or interval is defined"
      end
      if ( ! self[:interval].nil? && ! self[:ttl].nil? ) && ( self[:interval] != :absent && self[:ttl] != :absent )
        if self[:interval].to_i >= self[:ttl].to_i
          fail "check ttl #{self[:ttl]} must be greater than interval #{self[:interval]}"
        end
      end
    end
  end
end
